File: python/datastructures/arrayqueue.py
# arrayqueue.py

class EmptyQueueException(Exception):
    pass

class FullQueueException(Exception):
    pass

class Arrayqueue:
    def __init__(self, capacity):
        self.__data = [0] * capacity
        self.__front = 0
        self.__rear = 0
        self.__capacity = capacity
    
    def size(self):
        return (self.__capacity - self.__front + self.__rear) % self.__capacity
    
    def isempty(self):
        return self.size() == 0
    
    def front(self):
        if self.isempty():
            raise EmptyQueueException()
        return self.__data[self.__front]
    
    def enqueue(self, element):
        if self.size() == self.__capacity - 1:
            raise FullQueueException()
        self.__data[self.__rear] = element
        self.__rear = (self.__rear + 1) % self.__capacity
    
    def dequeue(self):
        if self.isempty():
            raise EmptyQueueException()
        element = self.__data[self.__front]
        self.__data[self.__front] = None
        self.__front = (self.__front + 1) % self.__capacity
        return element

File: csharp/Datastructures/ArrayQueue.cs
// ArrayQueue.cs

using System;

namespace Ukonline.Datastructures
{
	/// Array-based implementation of a queue.
	public class ArrayQueue<E> : Queue<E>
	{
		private readonly E[] data;      // Elements of this queue
		private int front, rear;        // Index of front and rear of this queue
		private readonly int capacity;  // Capacity of this queue
		private int size;               // Size of this stack

		/// <summary>
		/// Creates a new empty queue.
		/// </summary>
		/// <param name="capacity">The capacity of the queue.</param>
		public ArrayQueue (int capacity)
		{
			data = new E[capacity];
			front = 0;
			rear = 0;
			size = 0;
			this.capacity = capacity;
		}

		public int Size()
		{
			return size;
		}

		public bool IsEmpty()
		{
			return front == rear;
		}

		public E Front()
		{
			if (IsEmpty())
			{
				throw new EmptyQueueException();
			}
			return data[front];
		}

		public void Enqueue (E element)
		{
			if (Size() == capacity)
			{
				throw new FullQueueException();
			}
			data[rear] = element;
			rear = (rear + 1) % capacity;
			size++;
		}

		public E Dequeue()
		{
			if (IsEmpty())
			{
				throw new EmptyQueueException();
			}
			E element = data[front];
			data[front] = default(E);
			front = (front + 1) % capacity;
			size--;
			return element;
		}
	}
}